pkg/microservice/aslan/core/system/handler/sonar.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2022 The KodeRover Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handler
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/koderover/zadig/pkg/tool/crypto"
+	"io/ioutil"
+
+	"github.com/gin-gonic/gin"
+
+	commonservice "github.com/koderover/zadig/pkg/microservice/aslan/core/common/service"
+	"github.com/koderover/zadig/pkg/microservice/aslan/core/system/service"
+	internalhandler "github.com/koderover/zadig/pkg/shared/handler"
+	e "github.com/koderover/zadig/pkg/tool/errors"
+	"github.com/koderover/zadig/pkg/tool/log"
+)
+
+func CreateSonarIntegration(c *gin.Context) {
+	ctx := internalhandler.NewContext(c)
+	defer func() { internalhandler.JSONResponse(c, ctx) }()
+
+	args := new(service.SonarIntegration)
+
+	data, err := c.GetRawData()
+	if err != nil {
+		log.Errorf("Create sonar integration GetRawData err : %s", err)
+	}
+	if err = json.Unmarshal(data, args); err != nil {
+		log.Errorf("Create sonar integration Unmarshal err : %s", err)
+	}
+	internalhandler.InsertOperationLog(c, ctx.UserName, "", "新增", "系统配置-Sonar集成", fmt.Sprintf("server: %s, token: %s", args.ServerAddress, args.Token), string(data), ctx.Logger)
+
+	c.Request.Body = ioutil.NopCloser(bytes.NewBuffer(data))
+
+	if args.ServerAddress == "" || args.Token == "" {
+		ctx.Err = errors.New("name and server must be provided")
```

Comment:
Abstract this error into a public place?

Reply:
well this is not a common error, it is API specific